application/src/main/kotlin/net/corda/v5/application/flows/FlowContextProperties.kt

```diff
@@ -0,0 +1,45 @@
+package net.corda.v5.application.flows
+
+/**
+ * Context properties of a [Flow] are key value pairs of Strings. They are comprised of two sets of properties, those
+ * set by the platform when a [Flow] is started, and those which are added by the CorDapp developer during the execution
+ * of their flow. The latter set are referred to as user context properties. All context properties are immutable once
+ * set.
+ * Where keys exist as platform properties already, setting user properties is an effective no-o as they will always be
+ * ignored in preference for the platform value.
+ * Both sets of context properties are passed from the originating [Flow] to sub-flows, initiated flows, and services.
+ * Where sub-flows and initiated flows have extra user properties added, these are only visible in the scope of those
+ * flows and any of their sub-flows, initiated flows or services, but not back up the flow stack to any flows which
+ * sit above the one in which they were added. In this way context properties can also be thought of as a stack,
+ * unwinding along with the flow stack.
+ */
+interface FlowContextProperties {
+    /**
+     * Puts a user value into the context property store.
+     * If an existing user value exists anywhere in the current context stack, this value will overwrite it from this
+     * point on down, but not up, the context stack.
+     * If an existing platform value exists anywhere in the context stack, the value is ignored in preference for the
+     * platform value.
+     *
+     * @param key The property key
+     * @param value The property value
+     */
+    fun put(key: String, value: String)
+
+    /**
+     * Gets a value from the context property store.
+     * Also opens up [] operator access for getting values in Kotlin.
+     *
+     * @param key The property key
+     */
+    operator fun get(key: String): String?
+
+    /**
+     * Puts a user value into the context property store.
+     * Opens up [] operator access for setting values in Kotlin.
+     *
+     * @param key The property key
+     * @param value The property value
+     */
+    operator fun set(key: String, value: String)
```

Comment:
Do we need both `put` and `set`?

Reply:
set() is a Kotlin operator which gives you [] =, but put() makes it more Java Map like, the latter being inspired by what you said in last Friday's demo :-D

Reply:
Does Java end up seeing both?

Reply:
Yes, if it didn't Java users wouldn't be able to see get() either. So the only potential change is to remove the Kotlin-isms here I think. The is only a user facing class so I have no concerns that it will affect us internally if we make it less Kotlin like. I don't mind either way in other words, happy to go with consensus. I guess we don't hide the fact the api is really Kotlin made Java compatible generally, so I'm not sure if this is any more confusing that other parts of the api for Java users or not.

Reply:
You should be able to use an extension function to bypass this. For example:

```kotlin
interface FlowContextProperties {
    /**
     * Puts a user value into the context property store.
     * If an existing user value exists anywhere in the current context stack, this value will overwrite it from this
     * point on down, but not up, the context stack.
     * If an existing platform value exists anywhere in the context stack, the value is ignored in preference for the
     * platform value.
     *
     * @param key The property key
     * @param value The property value
     */
    fun put(key: String, value: String)

    /**
     * Gets a value from the context property store.
     * Also opens up [] operator access for getting values in Kotlin.
     *
     * @param key The property key
     */
    operator fun get(key: String): String?
}

/**
 * Puts a user value into the context property store.
 * Opens up [] operator access for setting values in Kotlin.
 *
 * @param key The property key
 * @param value The property value
 */
operator fun FlowContextProperties.set(key: String, value: String) = put(key, value)

class FlowContextPropertiesImpl : FlowContextProperties {
    override fun put(key: String, value: String) {
        TODO("Not yet implemented")
    }

    override fun get(key: String): String? {
        TODO("Not yet implemented")
    }
}

fun main() {
    val map = FlowContextPropertiesImpl()
    map[""] = ""
    map.put("", "")
    val a = map[""]
}
```

Extension functions are _harder_ to call from Java because they become static functions so there's no reason for Java users to call it when we use it like this. It ends up like this from Java:

```java
public class Test {

    public static void main(String[] args) {
        FlowContextProperties map = new FlowContextPropertiesImpl();
        // I defined FlowContextProperties in ExternalEventResponseFactoryKt
        ExternalEventResponseFactoryKt.set(map, "", "");
    }
}
```

At the end of the day its not a big deal, I just want to remove the possible confusion of "do I call put or set?".

Reply:
Yes, that's a good point, it also saves the effort of implementing it on the other side. I'll make this change.